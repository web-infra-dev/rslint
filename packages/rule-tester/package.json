{
  "name": "@typescript-eslint/rule-tester",
  "version": "0.1.4",
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "private": true,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "default": "./dist/"
    },
    "./package.json": "./package.json"
  },
  "scripts": {
    "build": "tsc -b tsconfig.build.json --force",
    "dev": "tsc -b tsconfig.build.json --force --watch",
    "test": "rstest run --testTimeout=0",
    "test:update": "rstest run -u --testTimeout=0"
  },
  "files": [
    "dist"
  ],
  "keywords": [
    "rslint",
    "rule-tester",
    "testing",
    "linter"
  ],
  "author": "",
  "license": "MIT",
  "description": "Rule testing utilities for rslint",
  "devDependencies": {
    "@types/node": "24.0.14",
    "typescript": "5.8.3"
  },
  "dependencies": {
    "@rslint/core": "workspace:*"
  },
  "publishConfig": {
    "access": "public"
  }
}
