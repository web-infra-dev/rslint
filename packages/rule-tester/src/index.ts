// Forked and modified from https://github.com/typescript-eslint/typescript-eslint/blob/16c344ec7d274ea542157e0f19682dd1930ab838/packages/rule-tester/src/RuleTester.ts#L4

import path from 'node:path';
import { test, describe, expect } from '@rstest/core';
import { lint, type Diagnostic } from '@rslint/core';
import assert from 'node:assert';

interface TsDiagnostic {
  line: number;
  column: number;
  endLine: number;
  endColumn: number;
  messageId: string;
  suggestions: any[];
}
function toCamelCase(name: string): string {
  return name.replace(/-([a-z])/g, g => g[1].toUpperCase());
}
// check whether rslint diagnostics and typescript-eslint diagnostics are semantic equal
function checkDiagnosticEqual(
  rslintDiagnostic: Diagnostic[],
  tsDiagnostic: TsDiagnostic[],
) {
  assert(
    rslintDiagnostic.length === tsDiagnostic.length,
    `Length mismatch: ${rslintDiagnostic.length} !== ${tsDiagnostic.length}`,
  );
  for (let i = 0; i < rslintDiagnostic.length; i++) {
    const rslintDiag = rslintDiagnostic[i];
    const tsDiag = tsDiagnostic[i];
    // check rule match
    assert(
      toCamelCase(rslintDiag.messageId) === tsDiag.messageId,
      `Message mismatch: ${rslintDiag.messageId} !== ${tsDiag.messageId}`,
    );

    // check range match
    // tsDiag sometimes doesn't have line and column, so we need to check that
    if (tsDiag.line) {
      assert(
        rslintDiag.range.start.line === tsDiag.line,
        `Start line mismatch: ${rslintDiag.range.start.line} !== ${tsDiag.line}`,
      );
    }
    if (tsDiag.endLine) {
      assert(
        rslintDiag.range.end.line === tsDiag.endLine,
        `End line mismatch: ${rslintDiag.range.end.line} !== ${tsDiag.endLine}`,
      );
    }
    if (tsDiag.column) {
      assert(
        rslintDiag.range.start.column === tsDiag.column,
        `Start column mismatch: ${rslintDiag.range.start.column} !== ${tsDiag.column}`,
      );
    }
    if (tsDiag.endColumn) {
      assert(
        rslintDiag.range.end.column === tsDiag.endColumn,
        `End column mismatch: ${rslintDiag.range.end.column} !== ${tsDiag.endColumn}`,
      );
    }
  }
}

interface RuleTesterOptions {
  languageOptions?: {
    parserOptions?: {
      project?: string;
      tsconfigRootDir?: string;
    };
  };
}
export class RuleTester {
  options: RuleTesterOptions;
  constructor(options: RuleTesterOptions) {
    this.options = options;
  }
  public run(
    ruleName: string,
    cases: {
      valid: (
        | string
        | { code: string; options?: any; only?: boolean; skip?: boolean }
      )[];
      invalid: {
        code: string;
        errors: any[];
        options?: any;
        only?: boolean;
        skip?: boolean;
      }[];
    },
  ) {
    describe(ruleName, () => {
      let cwd =
        this.options.languageOptions?.parserOptions?.tsconfigRootDir ||
        process.cwd();
      const config = path.resolve(cwd, './rslint.json');
      let virtual_entry = path.resolve(cwd, 'src/virtual.ts');
      // test whether case has only
      let hasOnly =
        cases.valid.some(x => {
          if (typeof x === 'object' && x.only) {
            return true;
          } else {
            return false;
          }
        }) || cases.invalid.some(x => x.only);
      test('valid', async () => {
        for (const validCase of cases.valid) {
          if (typeof validCase === 'object' && validCase.skip) {
            continue;
          }
          if (hasOnly) {
            if (typeof validCase === 'string') {
              continue;
            }
            if (!validCase.only) {
              continue;
            }
          }
          const code =
            typeof validCase === 'string' ? validCase : validCase.code;

          const options =
            typeof validCase === 'string' ? [] : validCase.options || [];

          // workaround for this hardcoded path https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/no-floating-promises.test.ts#L712
          if (Array.isArray(options)) {
            for (const opt of options) {
              if (Array.isArray(opt.allowForKnownSafeCalls)) {
                for (const item of opt.allowForKnownSafeCalls) {
                  if (item.path) {
                    item.path = virtual_entry;
                  }
                }
              }
            }
          }
          const diags = await lint({
            config,
            workingDirectory: cwd,
            fileContents: {
              [virtual_entry]: code,
            },
            ruleOptions: {
              [ruleName]: options,
            },
          });

          assert(
            diags.diagnostics?.length === 0,
            `Expected no diagnostics for valid case, but got: ${JSON.stringify(diags)}`,
          );
        }
      });
      test('invalid', async t => {
        for (const item of cases.invalid) {
          const {
            code,
            errors,
            only = false,
            skip = false,
            options = [],
          } = item;
          if (skip) {
            continue;
          }
          if (hasOnly && !only) {
            continue;
          }
          const diags = await lint({
            config,
            workingDirectory: cwd,
            fileContents: {
              [virtual_entry]: code,
            },
            ruleOptions: {
              [ruleName]: options,
            },
          });
          expect(diags).toMatchSnapshot();

          assert(
            diags.diagnostics?.length > 0,
            `Expected diagnostics for invalid case`,
          );
          checkDiagnosticEqual(diags.diagnostics, errors);
        }
      });
    });
  }
}

/**
 * Simple no-op tag to mark code samples as "should not format with prettier"
 *   for the plugin-test-formatting lint rule
 */
export function noFormat(raw: TemplateStringsArray, ...keys: string[]): string {
  return String.raw({ raw }, ...keys);
}
