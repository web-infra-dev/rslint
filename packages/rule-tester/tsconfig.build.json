{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "src",
    "noEmit": false
  },
  "include": ["src/**/*"],
  "exclude": ["dist", "node_modules"]
}
