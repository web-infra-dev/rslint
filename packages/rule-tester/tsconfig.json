{
  "extends": "../../tsconfig.spec.json",
  "include": ["src/**/*", "tests/**/*"],
  "exclude": ["dist", "node_modules"]
}
