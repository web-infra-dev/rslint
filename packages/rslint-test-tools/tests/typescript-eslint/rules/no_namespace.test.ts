import { noFormat, RuleTester } from '@typescript-eslint/rule-tester';

import { getFixturesRootDir } from '../RuleTester';

const rootPath = getFixturesRootDir();

const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootPath,
    },
  },
});

ruleTester.run('no-namespace', {
  valid: [
    // Regular module declaration (not namespace)
    `
      declare module "foo" {
        export const bar: string;
      }
    `,

    // Global module augmentation
    `
      declare global {
        interface Window {
          foo: string;
        }
      }
    `,

    // Ambient module declaration
    `
      declare module "bar" {
        export const baz: number;
      }
    `,

    // Declare namespace (allowed when allowDeclarations is true)
    {
      code: `
        declare namespace Test {
          export const value = 1;
        }
      `,
      options: [{ allowDeclarations: true }],
    },

    // Regular TypeScript code without namespaces
    `
      const value = 1;
      function test() {
        return value;
      }
      class Test {
        constructor() {}
      }
    `,

    // Module with exports (not namespace)
    `
      export const value = 1;
      export function test() {
        return value;
      }
    `,

    // Test array format options
    {
      code: `
        declare namespace Test {
          export const value = 1;
        }
      `,
      options: [{ allowDeclarations: true }],
    },

    // Test empty options object
    {
      code: `
        const value = 1;
      `,
      options: [{}],
    },

    // Test nil options
    `
      const value = 1;
    `,
  ],

  invalid: [
    // Basic namespace usage
    {
      code: `
        namespace Test {
          export const value = 1;
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Nested namespace
    {
      code: `
        namespace Outer {
          namespace Inner {
            export const value = 1;
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Namespace with interface
    {
      code: `
        namespace Test {
          export interface Config {
            value: string;
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Namespace with class
    {
      code: `
        namespace Test {
          export class MyClass {
            constructor() {}
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Namespace with function
    {
      code: `
        namespace Test {
          export function myFunction() {
            return "test";
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Declare namespace (not allowed by default)
    {
      code: `
        declare namespace Test {
          export const value = 1;
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Multiple namespaces
    {
      code: `
        namespace A {
          export const a = 1;
        }

        namespace B {
          export const b = 2;
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Namespace with complex content
    {
      code: `
        namespace Utils {
          export interface Options {
            debug?: boolean;
            timeout?: number;
          }

          export class Helper {
            static process(options: Options): void {
              // implementation
            }
          }

          export function validate(input: string): boolean {
            return input.length > 0;
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test allowDeclarations explicitly set to false
    {
      code: `
        declare namespace Test {
          export const value = 1;
        }
      `,
      options: [{ allowDeclarations: false }],
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test array options format but allowDeclarations false
    {
      code: `
        declare namespace Test {
          export const value = 1;
        }
      `,
      options: [{ allowDeclarations: false }],
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test mix of namespace and module declaration
    {
      code: `
        namespace Test {
          export const value = 1;
        }

        declare module "external" {
          export const externalValue = 2;
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test mix of namespace and global declaration
    {
      code: `
        namespace Test {
          export const value = 1;
        }

        declare global {
          interface GlobalInterface {
            prop: string;
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test deeply nested namespaces
    {
      code: `
        namespace Level1 {
          namespace Level2 {
            namespace Level3 {
              export const value = 1;
            }
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
        {
          messageId: 'noNamespace',
        },
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test namespace with type aliases
    {
      code: `
        namespace Types {
          export type StringOrNumber = string | number;
          export type Callback<T> = (value: T) => void;
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test namespace with enums
    {
      code: `
        namespace Constants {
          export enum Status {
            Active = "active",
            Inactive = "inactive"
          }
        }
      `,
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },

    // Test regular namespace should be reported even when allowDefinitionFiles is true
    {
      code: `
        namespace Test {
          export const value = 1;
        }
      `,
      options: [{ allowDefinitionFiles: true }],
      errors: [
        {
          messageId: 'noNamespace',
        },
      ],
    },
  ],
});
