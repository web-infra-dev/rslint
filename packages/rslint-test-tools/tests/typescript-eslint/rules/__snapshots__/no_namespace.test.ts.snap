// Rstest Snapshot v1

exports[`no-namespace > invalid 1`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 23,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 2`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 24,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 26,
          "line": 3,
        },
        "start": {
          "column": 21,
          "line": 3,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 2,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 3`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 23,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 4`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 23,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 5`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 23,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 6`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 31,
          "line": 2,
        },
        "start": {
          "column": 27,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 7`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 20,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 20,
          "line": 6,
        },
        "start": {
          "column": 19,
          "line": 6,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 2,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 8`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 24,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 9`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 31,
          "line": 2,
        },
        "start": {
          "column": 27,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 10`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 31,
          "line": 2,
        },
        "start": {
          "column": 27,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 11`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 23,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 12`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 23,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 13`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 25,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 27,
          "line": 3,
        },
        "start": {
          "column": 21,
          "line": 3,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 29,
          "line": 4,
        },
        "start": {
          "column": 23,
          "line": 4,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 3,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 14`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 24,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 15`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 28,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;

exports[`no-namespace > invalid 16`] = `
{
  "diagnostics": [
    {
      "filePath": "src/virtual.ts",
      "message": "Namespace is not allowed.",
      "messageId": "noNamespace",
      "range": {
        "end": {
          "column": 23,
          "line": 2,
        },
        "start": {
          "column": 19,
          "line": 2,
        },
      },
      "ruleName": "@typescript-eslint/no-namespace",
    },
  ],
  "errorCount": 1,
  "fileCount": 1,
  "ruleCount": 1,
}
`;
